### DISCUSSION TITLE
Are NDI Sources Cached?

### DISCUSSION THREAD

**Discussion by jasonmvp** [DistroAV/DistroAV#672] [Q&A - Support] [answered]
I have an NDI source, my Windows box, sitting on two different VLANs (public, private).  The latter of which is 10GigE.
My NDI client (Mac Pro 2019, running OBS Studio w/plugin) is also on the same two VLANs.

I'd like to force NDI to use the private VLAN but I'm uncertain how to do that.  I've tried blocking the Mac from talking to the Windows box over the public VLAN, but that just kills NDI.  It never attempts to connect over the private one.  It smells to me like there's a cache of some sort of known NDI sources, and if it can't hit them, it displays nothing.  But that's merely a guess.

Is the NDI plugin for OBS using some sort of cache?  Or is there some other mechanic at play here?

Thanks.

**Comment by jasonmvp** [comment, marked as answer]
It would seem that there is some sort of caching mechanism going on here.  I logged into my same Mac with another user, and set up the same scene w/OBS.  Before doing that, I made sure to block the Windows box from talking to the Mac over the public VLAN (using Mac's pf filter).  When I enabled the aforementioned NDI source, it popped right up in OBS.  And using tcpdump on the Mac, I was able to verify that it was coming over the private VLAN, not the public.

So: if there's a cache file, I assume it's in the user's ~/Library directory somewhere?  If so, where is it so I can delete it? :-)

**Reply by Heraes-git** [reply]
Did you find a way to clear the NDI cache ? I'm desperatly searching for a solution. My NDI plugin keeps using a NIC and trying to connect to a LAN PC that I have blocked with an ARP trick (similar to unpluging the cable). I even reinstalled the plugin and rebooted but it still connects to that god damn IP.

**Reply by jasonmvp** [reply]
> I even reinstalled the plugin and rebooted but it still connects to that god damn IP.

I never did figure out where the cache is stored.  The plugin just started using the correct interface after a while.  The connection is happening from the client side, not the source.  So if the PC is the client, your ARP trick might not work.  That is unless you're blocking the ARPs on both sides.

**Reply by Heraes-git** [reply]
Thanks for the clever insight. I've just tried to apply the same ARP blocking on the streaming PC ("client" of the NDI stream if I follow you), and rebooted it : it still doesn't work. :( I'm out of ideas. 
I see a connection between the two PC, in the good direction, but stuck in a waiting to a response ("SYN_SENT" code).

![2021-10-15_17h09m43s_SystemExplorer](https://user-images.githubusercontent.com/22033256/137529693-4b3a8709-a704-4475-9fe1-b2e45357c50e.jpg)
 
I can ping in both directions. I can also send a NDI stream from the streaming PC to the Gaming PC as a test, and it works. But it doesn't work in the other direction (the one I want, Jesus).

**Reply by Heraes-git** [reply]
**I successfully managed to clear the "cache" by doing a reboot of my internet box, and at the same time an `ipconfig /flushdns` on both PC.** I lost Internet access for a while, but by going in the box's interface (in the web browser), I saw that the Internet was cut. I made a software reboot (clicked a button) and everything went fine.
I'm gonna post one day the complete solution in public, because I think many people have the "black screen" problem.

**Reply by Heraes-git** [reply]
ERATTUM : this is fucking trying to contact the blocked NIC again... OMFG. From where does this retarded plugin find the adresse of a fucking card that is ARP banned ??? I can't even PING it ! Jesus...

**Reply by tt2468** [reply]
All networking-related functionality is handled by NDIlib. Unfortunately, there's not much we can do about its bad behavior.

**Comment by Heraes-git** [comment]
@Palakis Can you answer the question about where is the cache of the plugin ?

**Comment by Trouffman** [comment]
All network stuff are managed by the NDILib libraries directly, There is no caching in the plugin.

Note: the discovery process and the NDI stream are not passing through the same process/port.

More info on NDI internals : https://docs.ndi.video/docs/white-paper/introduction
